### PR TITLE
FastRandomContext improvements and switch to ChaCha20

### PR DESCRIPTION
**Pull request by sipa**
This switches FastRandomContext to use a ChaCha20-based random number generator. It also makes the class richer by adding support for getting single bits of entropy.

Benchmarks (also added) show that `rand32` became around 5.25x slower on my machine (from 1.5ns to 8ns), but the new `randbool` is 15% faster than the old one (1.3ns).

### CODE REVIEW

**Comment by gmaxwell** [comment]
src/addrman.cpp:                nKBucket = (nKBucket + insecure_rand.rand32()) % ADDRMAN_TRIED_BUCKET_COUNT;
src/addrman.cpp:                nKBucketPos = (nKBucketPos + insecure_rand.rand32()) % ADDRMAN_BUCKET_SIZE;
src/addrman.cpp:                nUBucket = (nUBucket + insecure_rand.rand32()) % ADDRMAN_NEW_BUCKET_COUNT;
src/addrman.cpp:                nUBucketPos = (nUBucketPos + insecure_rand.rand32()) % ADDRMAN_BUCKET_SIZE;
src/bench/checkqueue.cpp:            p.resize(insecure_rand.rand32() % (PREVECTOR_SIZE*2));
src/net.h:                vAddrToSend[insecure_rand.rand32() % vAddrToSend.size()] = _addr;


Usage wants a randrange a lot more than a rand32

File: src/test/crypto_tests.cpp [hunk 1]
```diff
@@ -439,4 +440,29 @@ BOOST_AUTO_TEST_CASE(aes_cbc_testvectors) {
                   "b2eb05e2c39be9fcda6c19078c6a9d1b3f461796d6b0d6b2e0c2a72b4d80e644");
 }
 
+BOOST_AUTO_TEST_CASE(chacha20_testvector)
```
Comment (jonasschnelli) [inline src/test/crypto_tests.cpp]: Maybe add all test vectors (only 5) from the IEFT draft specs: https://github.com/jonasschnelli/chacha20poly1305/blob/master/tests.c#L35

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by sipa** [comment]
Added `randrange` and the test vectors @jonasschnelli suggested.

File: src/test/prevector_tests.cpp [hunk 2]
```diff
@@ -183,11 +183,8 @@ class prevector_tester {
     }
 
     ~prevector_tester() {
-        BOOST_CHECK_MESSAGE(passed, "insecure_rand_Rz: "
```
Comment (TheBlueMatt) [inline src/test/prevector_tests.cpp]: Looks like this disabled a bunch of tests?

File: src/random.cpp [hunk 3]
```diff
@@ -177,15 +177,12 @@ uint256 GetRandHash()
...
-FastRandomContext::FastRandomContext(bool fDeterministic) : bytebuf_size(0), bitbuf_size(0)
+FastRandomContext::FastRandomContext(bool fDeterministic) : requires_seed(false), bytebuf_size(0), bitbuf_size(0)
 {
     static const unsigned char detkey[16] = {11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11};
     if (fDeterministic) {
         rng.SetKey(detkey, sizeof(detkey));
     } else {
-        unsigned char key[32];
-        GetRandBytes(key, sizeof(key));
-        rng.SetKey(key, sizeof(key));
```
Comment (TheBlueMatt) [inline src/random.cpp]: ...except you never check requires_seed to do the actual seeding?

File: configure.ac [hunk 4]
```diff
@@ -533,6 +533,9 @@ AC_CHECK_DECLS([bswap_16, bswap_32, bswap_64],,,
                  #include <byteswap.h>
                  #endif])
 
+AC_MSG_CHECKING(for __builtin_clzl)
```
Comment (TheBlueMatt) [inline configure.ac]: Hmm...is it definitely the case that all the compilers we support have this? Can we not have some fallback for those that do not?

File: src/random.h [hunk 5]
```diff
@@ -35,6 +35,7 @@ void GetStrongRandBytes(unsigned char* buf, int num);
  */
 class FastRandomContext {
 private:
+    bool requires_seed;
```
Comment (TheBlueMatt) [inline src/random.h]: ...except you never check requires_seed to do the actual seeding?

File: src/random.h [hunk 6]
```diff
@@ -81,6 +81,16 @@ class FastRandomContext {
         }
     }
 
+    unsigned long randrange(unsigned long range)
+    {
+        --range;
+        int bits = 8 * sizeof(long) - __builtin_clzl(range);
```
Comment (gmaxwell) [inline src/random.h]: CLZ is undefined for 0, a range of 1 is dumb but might be mechanically generated in some case.  At a minimum there should be a comment that range must be greater than 1.

File: configure.ac (hunk 4 above)
Comment (sipa) [inline configure.ac]: I don't know of any compilers that don't. I'd be happy to write a fallback if there is one that isn't.

File: src/test/prevector_tests.cpp (hunk 2 above)
Comment (sipa) [inline src/test/prevector_tests.cpp]: Oops, nice catch. Fixed.

File: src/random.h (hunk 5 above)
Comment (sipa) [inline src/random.h]: Fixed, and added a test to catch that.

File: configure.ac (hunk 4 above)
Comment (laanwj) [inline configure.ac]: I think MSVC doesn't.

**Comment by sipa** [comment]
Added a wrapper for __builtin_clzl, added unit tests, and fixed a few edge cases.

File: src/random.h (hunk 6 above)
Comment (sipa) [inline src/random.h]: Fixed.

File: src/random.h [hunk 7]
```diff
@@ -89,6 +90,16 @@ class FastRandomContext {
         }
     }
 
+    uint64_t randrange(uint64_t range)
```
Comment (gmaxwell) [inline src/random.h]: This needs a comment that points out that range returned will be [0..range) and that range _must not_ be zero.

Comment (sipa) [inline src/random.h]: Fixed.

**Comment by sipa** [comment]
Rebased.

**Comment by TheBlueMatt** [comment]
Looks good to me. I didnt re-verify the chacha code is correct, and dont know that the makefile changes are sane.

**Review by gmaxwell — APPROVED** [review]
utACK

**Comment by laanwj** [comment]
utACK 4fd2d2f

**Pull request merged**